common/socket/src/main/java/io/helidon/common/socket/IdleInputStream.java

```diff
@@ -36,14 +40,23 @@
  */
 class IdleInputStream extends InputStream {
 
+    private static final int TIMEOUT = 500;
+
+    private final Socket socket;
     private final InputStream upstream;
     private final LazyValue<ExecutorService> executor;
+    private final int idleTimeoutIterations;
+    private final int lastIterationRemainder;
     private volatile int next = -1;
     private volatile boolean closed = false;
+    private volatile boolean cancelled = false;
     private Future<?> idlingThread;
 
-    IdleInputStream(InputStream upstream, String childSocketId, String socketId) {
+    IdleInputStream(Socket socket, InputStream upstream, String childSocketId, String socketId, Duration idleTimout) {
+        this.socket = socket;
         this.upstream = upstream;
+        this.idleTimeoutIterations = (int) (idleTimout.toMillis() / TIMEOUT);
+        this.lastIterationRemainder = (int) (idleTimout.toMillis() % TIMEOUT);
```

Comment:
After looking at the final version of this, I think there's no need for the "remainder" logic TBH. Timeouts are approximations (especially at the socket layer), so it should be enough to divide by TIMEOUT or maybe add 1 to that. No need for finer control than that. That should simplify your logic a bit, making it easier to read.

Reply:
+1 the loop should just increment the max iterations

Reply:
Something like this maybe ?

```java
try {
    int timeout = socket.getSoTimeout();
    for (int i = 0; i < maxIterations && !cancelled; i++) {
        try {
            int t = socket.getSoTimeout();
            if (t != 500) {
                timeout = t;
            }
            socket.setSoTimeout(500);
            next = upstream.read();
            if (next <= 0) {
                closed = true;
            }
            break;
        } catch (SocketTimeoutException ignored) {
        }
    }
    socket.setSoTimeout(timeout);
} catch (IOException e) {
    closed = true;
    throw new UncheckedIOException(e);
}
```

Reply:
Alright, I have simplified the logic.